entity-framework/core/querying/related-data.md

```diff
@@ -47,8 +47,39 @@ You may want to include multiple related entities for one of the entities that i
 
 [!code-csharp[Main](../../../samples/core/Querying/RelatedData/Sample.cs#MultipleLeafIncludes)]
 
-> [!CAUTION]
-> Since version 3.0.0, each `Include` will cause an additional JOIN to be added to SQL queries produced by relational providers, whereas previous versions generated additional SQL queries. This can significantly change the performance of your queries, for better or worse. In particular, LINQ queries with an exceedingly high number of `Include` operators may need to be broken down into multiple separate LINQ queries in order to avoid the cartesian explosion problem.
+### Avoiding cartesian explosion
+
+> [!NOTE]
+> The behavior described below is new since version 3.0.0, and is only relevant for relational databases.
+
+When targeting a relational database, `Include` and `ThenInclude` operators cause an additional JOIN to be added to your SQL query. A large number of JOINs can result in a so-called *cartesian explosion*, where data is returned multiple times from the database, causing potentially serious performance issues.
+
+For example, consider the following query:
+
+``` csharp
+var customers = context.Customers
+    .Include(c => c.Addresses)
+    .Include(c => c.Orders)
+        .ThenInclude(o => o.OrderLine)
+    .Where(c => c.CustomerName == "John")
+    .ToList();
+```
+
+Because of the way relational databases work, a customer with several addresses and/or orders will be duplicated in the database resultset; address fields will be duplicated as well, based on the number of orders and order details the same customer has, and so on. This may not be significant if the number of `Include` operators is low or if the database contains little data, but as additional collection navigations are included duplication quickly grows.
+
+The solution is to split this query into several smaller queries using [explicit loading](#explicit-loading), where each query includes a smaller number of collection navigations (possibly just one):
+
+``` csharp
+var baseQuery = db.Customers.Where(c => c.CustomerName == "John");
+var customers = baseQuery.ToList();  // Loads customers without any related entities
+baseQuery.SelectMany(c => c.Addresses).Load();  // Loads only the addresses
```

Comment:
This is not explicit loading.

Reply:
How is it not [explicit loading](https://docs.microsoft.com/en-us/ef/core/querying/related-data#explicit-loading):

> You can also explicitly load a navigation property by executing a separate query that returns the related entities.

Reply:
Explicit load APIs are defined on NavigationEntry. This is not navigation entry. This is pure IQueryable. Load function here is whole lot different than explicit loading.

Reply:
OK, I guess it's terminological but the functional idea and outcome are the same: load related entities explicitly after loading the principal - that's even how our docs say it. Whether it's done on NavigationEntry or on the IQueryable doesn't seem very important (at least to users). But I don't care that much either way.

Reply:
Outcome are altogether different. Explicit load API builds query with where predicate on key and can load only 1 entity's navigation at a time. You cannot even load reference navigation for items in the collection with that. At best using explicit loading for collection include gives you N + 1 queries and not 2 queries.